### ISSUE TITLE
[request] TensorBoard logger should not be (a hyperparameter) in LightningModule

### ISSUE THREAD

**Issue by bencomp** [mittagessen/kraken#477]
While writing, I may have thought of a workaround. Still, I think this request could be considered and if it is, I can probably create a PR for it. I think the core of my issue is that the TensorBoard logger (info) is given to both the `RecognitionModel` (and `SegmentationModel`) as well as the `KrakenTrainer`, whereas IMHO it should only be a concern of the `KrakenTrainer`.

I would like to train a recognition model with monitoring using TensorBoard and specify extra callbacks, like a [`DeviceStatsMonitor`][dsm]. I also want to specify the name of the model as the `name` argument of the [`TensorBoardLogger`][tbl].
To prevent modifying Kraken's code I copied and modified the `ketos` CLI, which creates the `RecognitionModel` and `KrakenTrainer` with a few parameters from [`pl.Trainer`][plt].

[dsm]: https://lightning.ai/docs/pytorch/stable/api/pytorch_lightning.callbacks.DeviceStatsMonitor.html#pytorch_lightning.callbacks.DeviceStatsMonitor
[tbl]: https://lightning.ai/docs/pytorch/stable/extensions/generated/lightning.pytorch.loggers.TensorBoardLogger.html#lightning.pytorch.loggers.TensorBoardLogger
[plt]: https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#trainer-class-api

Based on https://github.com/mittagessen/kraken/blob/44a9933266a345f8ba1013f4ca5de3ce14f20808/kraken/lib/train.py#L61-L76 I figured I could overwrite the `'tensorboard'` value of `pl_logger` with my own `TensorBoardLogger` with the right parameters and provide it to the `KrakenTrainer` as the `pl_logger` and the `logger` *kwarg*.

However, `pl_logger` in the `ketos train` command is also saved as a hyperparameter and since a `TensorBoardLogger` is not JSON-encodable, the training crashed when the model was to be saved.

A simple workaround appeared to be to remove `pl_logger` from the hyperparameters provided to the `RecognitionModel`. But then I could not monitor the learning rate, as the `LearningRateMonitor` is only added if `pl_logger` is a hyperparameter:
https://github.com/mittagessen/kraken/blob/44a9933266a345f8ba1013f4ca5de3ce14f20808/kraken/lib/train.py#L575-L584

I would like to propose:

- remove `pl_logger` from the training hyperparameters
- add the `LearningRateMonitor` in `KrakenTrainer.__init__`
- optionally add CLI options for the experiment name, logging the graph and creating a `DeviceStatsMonitor` callback

How do others feel about this?

**Comment by colibrisson** [comment]
I agree: `pl_logger` should not be in `hparams`. I think I saw that `EarlyStopping` was at the model level and I just followed along. I apologize for that ;-)

**Comment by bencomp** [comment]
Thanks for your response, @colibrisson :) I'm not here to blame anyone – if it felt that way, please accept my apologies. I am glad that you agree. Let me start work on moving some TB logger things to better locations.

From how I understand PyTorch Lightning, I think it makes sense that `EarlyStopping` is configured in the module, although I could also be convinced that it should be a callback on the `Trainer`.

**Comment by bencomp** [comment]
The PR was closed when the name of the master branch changed to main, but https://github.com/mittagessen/kraken/commit/10509a5d7b7ba7af6fe17fd182aa4516bacf1645 merged the suggested changes instead.
